# HG changeset patch
# User Bill Denny
# Date 1206645473 14400
#      Thu Mar 27 15:17:53 2008 -0400
# Node ID 4ab2488ab2b4c0be1f62c47f09279170e36de619
# Parent  1ce6460aebdf14e29463e2192d2483d8935fc25b
datenum: allow vector inputs in any orientation

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,10 +1,12 @@
 2008-03-27  Bill Denney  <bill@denney.ws>
 
+	* time/datenum.m: Allow vector inputs in any orientation.
+
 	* strings/validatestring.m: New function.
 	* strings/Makefile.in (SOURCES): Add it to the list.
 
 	* general/nargoutchk.m: New function.
 	* general/Makefile.in (SOURCES): Add it to the list.
 
 	* general/genvarname.m: New function.
 	* general/Makefile.in (SOURCES): Add it to the list.
diff --git a/scripts/time/datenum.m b/scripts/time/datenum.m
--- a/scripts/time/datenum.m
+++ b/scripts/time/datenum.m
@@ -1,9 +1,9 @@
-## Copyright (C) 2006, 2007 Paul Kienzle
+## Copyright (C) 2006, 2007, 2008 Paul Kienzle
 ##
 ## This file is part of Octave.
 ##
 ## Octave is free software; you can redistribute it and/or modify it
 ## under the terms of the GNU General Public License as published by
 ## the Free Software Foundation; either version 3 of the License, or (at
 ## your option) any later version.
 ##
@@ -94,17 +94,17 @@ function [days, secs] = datenum (Y, M, D
 
   M(M<1) = 1; ## For compatibility.  Otherwise allow negative months.
 
   ## Set start of year to March by moving Jan. and Feb. to previous year.
   ## Correct for months > 12 by moving to subsequent years.
   Y += fix ((M-14)/12);
 
   ## Lookup number of days since start of the current year.
-  D += monthstart (mod (M-1,12) + 1) + 60;
+  D += reshape (monthstart (mod (M-1,12) + 1), size (D)) + 60;
 
   ## Add number of days to the start of the current year. Correct
   ## for leap year every 4 years except centuries not divisible by 400.
   D += 365*Y + floor (Y/4) - floor (Y/100) + floor (Y/400);
 
   ## Add fraction representing current second of the day.
   days = D + (h+(m+s/60)/60)/24;
 
@@ -115,12 +115,18 @@ endfunction
 
 %!shared part
 %! part = 0.514623842592593;
 %!assert(datenum(2001,5,19), 730990)
 %!assert(datenum([1417,6,12]), 517712)
 %!assert(datenum([2001,5,19;1417,6,12]), [730990;517712])
 %!assert(datenum(2001,5,19,12,21,3.5), 730990+part, eps)
 %!assert(datenum([1417,6,12,12,21,3.5]), 517712+part, eps)
+## Test vector inputs
 %!test
 %! t = [2001,5,19,12,21,3.5; 1417,6,12,12,21,3.5];
 %! n = [730990; 517712] + part;
 %! assert(datenum(t), n, 2*eps);
+## Make sure that the vectors can have either orientation
+%!test
+%! t = [2001,5,19,12,21,3.5; 1417,6,12,12,21,3.5]';
+%! n = [730990 517712] + part;
+%! assert(datenum(t(1,:), t(2,:), t(3,:), t(4,:), t(5,:), t(6,:)), n, 2*eps);
